protocols/relay/examples/relay.rs

```diff
@@ -82,3 +154,149 @@ fn main() -> Result<(), Box<dyn Error>> {
+            DEFAULT_RELAY_ADDRESS.to_string()
+        }
+    }
+}
+
+/// Get the address for client_listen mode
+fn get_relay_peer_address(opt: &Opt) -> String {
+    match &opt.address {
+        Some(address) => address.clone(),
+        None => {
+            match get_shared_data(RELAY_ADDRESS_KEY) {
+                Some(address) => address,
+                None => panic!("Please provide relayed listen address such as: <addr-relay-server>/p2p/<peer-id-relay-server>/p2p-circuit"),
+            }
+        }
+    }
+}
+
+/// Get the address for client-dial mode
+fn get_client_listen_address(opt: &Opt) -> String {
+    match &opt.address {
+        Some(address) => address.clone(),
+        None => {
+            match get_shared_data(CLIENT_LISTEN_ADDRESS_KEY) {
+                Some(address) => address.to_string(),
+                None => panic!("Please provide client listen address such as: <addr-relay-server>/p2p/<peer-id-relay-server>/p2p-circuit/p2p/<peer-id-listening-relay-client>"),
+            }
+        }
+    }
+}
+
+/// Placeholder to publish the peer address based on given key
+fn set_shared_data(key: &str, value: &str) -> () {
+    println!(
+        "TODO publish to KAD or discovery service ... Set the key: {} data to:{}",
+        key, value
+    );
+}
+
+/// Placeholder to get the peer information based on given key
+fn get_shared_data(key: &str) -> Option<String> {
+    println!(
+        "TODO get the data related to key: {} from KAD or discovery service.",
+        key
+    );
+    None
+}
+
+#[derive(NetworkBehaviour)]
+#[behaviour(out_event = "Event", event_process = false)]
+struct Behaviour {
+    relay: Relay,
+    ping: Ping,
+}
+
+#[derive(Debug)]
+enum Event {
+    Relay(()),
+    Ping(PingEvent),
+}
+
+impl From<PingEvent> for Event {
+    fn from(e: PingEvent) -> Self {
+        Event::Ping(e)
+    }
+}
+
+impl From<()> for Event {
+    fn from(_: ()) -> Self {
+        Event::Relay(())
+    }
+}
+
+#[derive(Debug, StructOpt)]
+enum Mode {
+    Relay,
+    ClientListen,
+    ClientDial,
+}
+
+impl FromStr for Mode {
+    type Err = ModeError;
+    fn from_str(mode: &str) -> Result<Self, Self::Err> {
+        match mode {
+            "relay" => Ok(Mode::Relay),
+            "client-listen" => Ok(Mode::ClientListen),
+            "client-dial" => Ok(Mode::ClientDial),
+            _ => Err(ModeError {}),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ModeError {}
+impl Error for ModeError {}
+impl fmt::Display for ModeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Could not parse a mode")
+    }
+}
+
+#[derive(Debug, StructOpt)]
+#[structopt(name = "libp2p relay")]
+struct Opt {
+    /// The mode (relay, client-listen, client-dial)
+    #[structopt(long)]
+    mode: Mode,
+
+    /// Fixed value to generate deterministic peer id
+    #[structopt(short = "d")]
+    deterministic_peer_secret: u8,
```

Comment:
```suggestion
    secret_key_seed: u8,
```

The word `seed` would be more intuitive to me here.